snow/engine/snowman/event/queue_test.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package event
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	errDuplicateInvocation  = errors.New("job already handled")
+	errUnexpectedInvocation = errors.New("job handled unexpectedly")
+)
+
+type testJob struct {
+	execute func(context.Context) error
+	cancel  func(context.Context) error
+}
+
+func (j *testJob) Execute(ctx context.Context) error {
+	return j.execute(ctx)
+}
+
+func (j *testJob) Cancel(ctx context.Context) error {
+	return j.cancel(ctx)
+}
+
+func newQueueWithJob[T comparable](t *testing.T, job Job, shouldCancel bool, dependencies ...T) *Queue[T] {
+	q := NewQueue[T]()
+	require.NoError(t, q.Register(context.Background(), job, dependencies...))
+	if shouldCancel {
+		for _, jobs := range q.jobs {
+			for _, j := range jobs {
+				j.shouldCancel = true
+			}
+		}
+	}
+	return q
+}
+
+func TestQueue_Register(t *testing.T) {
+	var calledExecute bool
+	userJob := &testJob{
+		execute: func(context.Context) error {
+			if calledExecute {
+				return errDuplicateInvocation
+			}
+			calledExecute = true
+			return nil
+		},
+		cancel: func(context.Context) error {
+			return errUnexpectedInvocation
+		},
+	}
+
+	tests := []struct {
+		name          string
+		queue         *Queue[int]
+		dependencies  []int
+		shouldExecute bool
+		expectedLen   int
+		expectedQueue *Queue[int]
+	}{
+		{
+			name:          "no dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  nil,
+			shouldExecute: true,
+			expectedLen:   0,
+			expectedQueue: NewQueue[int](),
+		},
+		{
+			name:          "one dependency",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1},
+			shouldExecute: false,
+			expectedLen:   1,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:          "two dependencies",
+			queue:         NewQueue[int](),
+			dependencies:  []int{1, 2},
+			shouldExecute: false,
+			expectedLen:   2,
+			expectedQueue: &Queue[int]{
+				jobs: map[int][]*job[int]{
+					1: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+					2: {
+						{
+							dependencies: set.Of(1, 2),
+							job:          userJob,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "additional dependency",
+			queue: &Queue[int]{
```

Comment:
In the other test you use `newQueueWithJob(userJob)`; is that appropriate here too?

Reply:
I had avoided using it in this test because it internally calls `Register`... Which is what this function is testing... But I think it it's fine to use for the setup (not the expected).